Feat/time deal review

bjcareer opened the pull request:
타임딜의 서비스 구조는 아래와 같습니다.

## 순서도 
- 사용자의 발급 요청 수신: 사용자가 쿠폰 발급을 요청합니다.
- 이벤트 확인: DB에 접근하여 사용자가 요청한 이벤트가 실제로 존재하는지 확인합니다.
- 이벤트가 존재하지 않는 경우: 사용자에게 Bad Request 응답을 보냅니다.
- 이벤트가 존재하는 경우: 사용자가 해당 이벤트에서 쿠폰을 발급받은 이력이 있는지 확인합니다.
- 쿠폰 발급 이력 확인: 사용자가 해당 이벤트에서 쿠폰을 발급받은 이력이 있는지 확인합니다.
- 이력이 있는 경우: 사용자에게 이미 쿠폰을 발급받았다는 응답
- 
<img width="1920" alt="image" src="https://github.com/user-attachments/assets/ca294916-137d-46da-b9ab-bcb901ed861d">




## ERD
서비스 ERD (Entity-Relationship Diagram)에 대해 간단히 설명드리겠습니다. 현재 정의된 쿠폰 서비스는 다음과 같은 테이블들로 구성되어 있습니다:

### 테이블: Coupon
- **할인 금액 (discount_amount)**: 쿠폰을 사용하여 적용될 할인 금액을 나타냅니다.
- **발급 일자 (issued_at)**: 쿠폰이 발급된 날짜를 기록합니다.
- **쿠폰 번호 (coupon_code)**: UUID V4 형식으로 생성된 고유한 쿠폰 번호를 저장합니다.
- **사용 유무 (is_used)**: 쿠폰이 사용되었는지 여부를 나타내는 boolean 타입의 값입니다.

### 테이블: Time_Deal_Event
- **발급된 쿠폰의 개수 (total_coupons_issued)**: 이벤트를 통해 발급된 전체 쿠폰의 수를 기록합니다.
- **사용자에게 전달된 쿠폰 개수 (coupons_distributed_to_users)**: 각 사용자에게 몇 개의 쿠폰이 전달되었는지를 기록합니다.

### 관계
- **Time_Deal_Event (1) -> Coupon (N)**: 하나의 Time_Deal_Event가 여러 개의 쿠폰을 발급할 수 있는 일대다 관계입니다.

<img width="666" alt="스크린샷 2024-07-22 오후 5 05 58" src="https://github.com/user-attachments/assets/2b8adc9c-a500-47de-8f10-5f7055946288">

bjcareer commented:
[#](https://github.com/f-lab-edu/hot-stock/issues/8)

bjcareer committed:
[feat] Use Redis

- 레디스에 락을 이용해서 경쟁조건을 제거했습니다.
- 기존에는 쿠폰 발급을 위해서 DB에 락을 걸고 조회했다면, 변경된 코드는 레디스의 값을 조회하도록 변경했습니다.(key는 eventId, value는 TimeDealEventDomain)
- 디비에 저장하는 방식들은 모두 async 방식으로 변경하여 처리량 개선을 시도했습니다

related to #11

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/domain/RedisLock.java:
```diff
@@ -0,0 +1,41 @@
+        boolean result = false;
+
+        try {
+            result = lock.tryLock(1, 1, TimeUnit.MINUTES);
```
1과 같은 magic number등은 의도를 명확하게 하기 위해 상수 또는 설정파일을 통해 가져오면 좋습니다

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/repository/event/RedisTimeDealRepository.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public Long save(TimeDealEvent timeDealEvent) {
+        RBucket<TimeDealEvent> bucket = redissonClient.getBucket("timeDeal:" + timeDealEvent.getId());
```
Bucket이름에 사용된 string도 상수처리가 좋습니다.

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/domain/RedisLock.java:
```diff
@@ -0,0 +1,41 @@
+            result = lock.tryLock(1, 1, TimeUnit.MINUTES);
+        } catch (InterruptedException e) {
+            log.error(e.getMessage());
+            throw new IllegalStateException("Lock 요청 실패");
```
다음 사항을 고려해보면 좋습니다. 

- 재시도를 해볼 수 있는지
- exception을 발생시키는 대신 result를 리턴하는 방법

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/service/TimeDealService.java:
```diff
@@ -1,67 +1,90 @@
+        boolean isLocked = redisLock.tryLock(lockName);
+
+        if (isLocked){
+            Long saveId = timeDealEventRepository.save(timeDealEvent);
```
두개의 save작업 중 예외가 발생할 수 있는지, 있다면 어떤 방식으로 처리할지 고려해보시면 좋습니다.

bjcareer replied:
좋은 리뷰 감사합니다.
제가 이 부분을 확인해본 결과, 레디스 서버가 다운되거나 DB 서버가 다운된 상태가 아니라면 오류가 발생하지 않는 것을 확인했습니다. 혹시 서버가 다운됐을 때도 에러 처리를 따로 해줘야 할까요?

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/service/TimeDealService.java:
```diff
@@ -1,67 +1,90 @@
-        log.debug("사용자에게 전달된 쿠폰의 개수는 = {} ", timeDealEvent.getDeliveredCouponNum());
-        vaildateRemainCoupon(timeDealEvent);
+        String lockName = "generateCouponToUserLock";
+        boolean is_locked = redisLock.tryLock(lockName);
```
Java naming convention을 따르는 것이 좋습니다.

f-lab-troy commented on timeDeal/src/main/java/com/bjcareer/stockservice/timeDeal/service/TimeDealService.java:
```diff
@@ -1,67 +1,90 @@
+        String lockName = "generateCouponToUserLock";
+        boolean is_locked = redisLock.tryLock(lockName);
 
+        if (!is_locked) {
```
위 다른 코드에서 isLocked를 체크했기때문에 일관성을 위해 이곳에서 반대조건을 체크하는 것은 피하는 것이 좋을 수 있습니다.

f-lab-troy commented on timeDeal/src/test/java/com/bjcareer/stockservice/timeDeal/service/TimeDealServiceTest.java:
```diff
@@ -2,33 +2,43 @@
 class TimeDealServiceTest {
     @Autowired private TimeDealService timeDealService;
     @Autowired private EventRepository timeDealEventRepository;
+    @Autowired private InMemoryEventRepository inMemoryEventRepository;
```
단위 테스트에서 실제 의존하는 class보다 mock을 사용하여 원하는 결과를 조정(예: 예외 상황)하는 것을 추천드립니다.

f-lab-troy commented on timeDeal/src/test/java/com/bjcareer/stockservice/timeDeal/service/TimeDealServiceTest.java:
```diff
@@ -72,9 +87,25 @@ void setUp() {
         thread.join();
         thread1.join();
 
+        Thread.sleep(500);
```
테스트에 sleep은 사용안하시는 것이 좋습니다.

bjcareer committed:
[refactor] Constantize magic number
 - 락을 획득하는데 기다리는 시간을 최대 1분으로 설정하였습니다.
 - 서비스 호출이 끝나면 사용자가 락을 반환함으로써 LEASE_TIME을 0으로 두어 비동기로 LOCK을 해제하는 코드가 동작하지 않게 했습니다.
 - 의도가 명확하지 않은 숫자들, 문자들을 상수로 변경했슴니다.

bjcareer committed:
[refactor] Update coupon generation logic and add custom exception handling

 - Controller에 예외를 올려주기 보단 Optional을 통해서 status code를 처리할 수 있게 변경했습니다.

 - generateCouponToUser의 경우에는 Redis lock을 획득하지 못하는 서버 오류, 잘못된 이벤트를 요청하는 사용자 오류를 나누기 위해서 error를 raise하는 방식을 택했습니다

 - 또한 리뷰에서 save할 때 발생할 수 있는 오류에 대해서 알아봤는데, Redis서버 종료 오류, DB 서버 종료 오류가 있었습니다.

bjcareer committed:
[refactor] Use mock
 - Mock을 이용해서 테스트하도록 변경
